File: python/aps/anchor_ethereum.py
"""Ethereum (EVM) anchor provider using raw JSON-RPC calls."""
from __future__ import annotations

import json
import urllib.request
from dataclasses import dataclass
from datetime import datetime, timezone

from .anchor import AnchorMetadata, AnchorProvider, AnchorReceipt, AnchorVerification, ProviderInfo

# anchor(bytes32) selector
ANCHOR_SELECTOR = "0xc2b12a73"
# isAnchored(bytes32) selector
IS_ANCHORED_SELECTOR = "0xa85f7489"


@dataclass
class EthereumConfig:
    rpc_url: str
    contract_address: str
    chain_id: str = "1"
    from_address: str = ""
    private_key: str = ""


class EthereumAnchor(AnchorProvider):
    def __init__(self, cfg: EthereumConfig, *, opener: urllib.request.OpenerDirector | None = None):
        self.cfg = cfg
        self._opener = opener or urllib.request.build_opener()

    def _rpc_call(self, method: str, params: list) -> object:
        body = json.dumps({"jsonrpc": "2.0", "method": method, "params": params, "id": 1}).encode()
        req = urllib.request.Request(self.cfg.rpc_url, data=body, headers={"Content-Type": "application/json"})
        resp = self._opener.open(req)
        data = json.loads(resp.read())
        if data.get("error"):
            raise RuntimeError(f"RPC error: {data['error']}")
        return data.get("result")

    def commit(self, hash_bytes: bytes, meta: AnchorMetadata) -> AnchorReceipt:
        data = ANCHOR_SELECTOR + hash_bytes.hex().ljust(64, "0")[:64]
        tx = {"from": self.cfg.from_address, "to": self.cfg.contract_address, "data": data}
        tx_hash = self._rpc_call("eth_sendTransaction", [tx])

        block = 0
        try:
            receipt = self._rpc_call("eth_getTransactionReceipt", [tx_hash])
            if receipt and receipt.get("blockNumber"):
                block = int(receipt["blockNumber"], 16)
        except Exception:
            pass

        return AnchorReceipt(
            tx_hash=tx_hash,
            block=block,
            timestamp=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
            provider="ethereum",
        )

    def verify(self, hash_bytes: bytes) -> AnchorVerification:
        data = IS_ANCHORED_SELECTOR + hash_bytes.hex().ljust(64, "0")[:64]
        call = {"to": self.cfg.contract_address, "data": data}
        result = self._rpc_call("eth_call", [call, "latest"])

        is_anchored = False
        if isinstance(result, str):
            clean = result.replace("0x", "")
            if len(clean) >= 64:
                is_anchored = clean[63] == "1"

        if not is_anchored:
            return AnchorVerification(exists=False)

        return AnchorVerification(
            exists=True,
            timestamp=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        )

    def info(self) -> ProviderInfo:
        return ProviderInfo(name="ethereum", chain_id=self.cfg.chain_id, type="ethereum")

File: typescript/src/anchor_ethereum.ts
import type { AnchorReceipt, AnchorVerification, ProviderInfo, AnchorMetadata } from './types.js';
import type { AnchorProvider } from './anchor.js';
import { bytesToHex } from '@noble/hashes/utils';

export interface EthereumConfig {
  rpcUrl: string;
  contractAddress: string;
  chainId?: string;
  fromAddress?: string;
  privateKey?: string;
}

// anchor(bytes32) selector
const ANCHOR_SELECTOR = '0xc2b12a73';
// isAnchored(bytes32) selector
const IS_ANCHORED_SELECTOR = '0xa85f7489';

interface JsonRpcResponse {
  jsonrpc: string;
  result?: unknown;
  error?: { code: number; message: string };
  id: number;
}

export class EthereumAnchor implements AnchorProvider {
  private cfg: EthereumConfig;
  private _fetch: typeof globalThis.fetch;

  constructor(cfg: EthereumConfig, fetchFn?: typeof globalThis.fetch) {
    this.cfg = cfg;
    this._fetch = fetchFn || globalThis.fetch;
  }

  private async rpcCall(method: string, params: unknown[]): Promise<unknown> {
    const body = JSON.stringify({ jsonrpc: '2.0', method, params, id: 1 });
    const resp = await this._fetch(this.cfg.rpcUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body,
    });
    const data: JsonRpcResponse = await resp.json();
    if (data.error) throw new Error(`RPC error: ${data.error.message}`);
    return data.result;
  }

  async commit(hash: Uint8Array, _meta: AnchorMetadata): Promise<AnchorReceipt> {
    const hashHex = bytesToHex(hash);
    const data = ANCHOR_SELECTOR + hashHex.padEnd(64, '0').slice(0, 64);
    const tx = { from: this.cfg.fromAddress || '', to: this.cfg.contractAddress, data };
    const txHash = (await this.rpcCall('eth_sendTransaction', [tx])) as string;

    let block = 0;
    try {
      const receipt = (await this.rpcCall('eth_getTransactionReceipt', [txHash])) as {
        blockNumber?: string;
      } | null;
      if (receipt?.blockNumber) {
        block = parseInt(receipt.blockNumber, 16);
      }
    } catch {
      // ignore
    }

    return {
      tx_hash: txHash,
      block,
      timestamp: new Date().toISOString(),
      provider: 'ethereum',
    };
  }

  async verify(hash: Uint8Array): Promise<AnchorVerification> {
    const hashHex = bytesToHex(hash);
    const data = IS_ANCHORED_SELECTOR + hashHex.padEnd(64, '0').slice(0, 64);
    const call = { to: this.cfg.contractAddress, data };
    const result = (await this.rpcCall('eth_call', [call, 'latest'])) as string;

    const clean = result.replace('0x', '');
    const isAnchored = clean.length >= 64 && clean[63] === '1';

    if (!isAnchored) return { exists: false };
    return { exists: true, timestamp: new Date().toISOString() };
  }

  info(): ProviderInfo {
    return { name: 'ethereum', chain_id: this.cfg.chainId || '1', type: 'ethereum' };
  }
}
